include/oneapi/dpl/pstl/iterator_impl.h

```diff
@@ -139,6 +139,110 @@ class zip_forward_iterator
     __it_types __my_it_;
 };
 
+template <typename T, typename _UnaryFunc>
+auto
+make_transform_output_ref_wrapper(T&& __reference, _UnaryFunc __unary_func);
+
+template <typename T, typename _UnaryFunc>
+class transform_output_ref_wrapper
+{
+  private:
+    T& __my_reference_;
+    _UnaryFunc __my_unary_func_;
+
+  public:
+    typedef ::std::true_type IsTransformOutputIterRefWrapper;
+    transform_output_ref_wrapper(T& __reference, _UnaryFunc __unary_func)
+        : __my_reference_(__reference), __my_unary_func_(__unary_func)
+    {
+    }
+
+    operator T&() { return __my_reference_; }
```

Comment:
Up to me - it's a little bit strange pattern.
Usually we have operator*(), like for example described in https://en.cppreference.com/w/cpp/iterator/iterator
I mean
reference operator*() const {return num;}

Is this operator for implicit conversion really required?

Reply:
ok, about transform_output_ref_wrapper it's my error, sorry.
But may be better way to define operator[] ?

Reply:
No problem, the code is definitely a abnormal pattern because it is trying to do something which is somewhat abnormal: have the write operation trigger a transform side effect.  It seems I need more comments to make this more maintainable.  

For the `operator[]` question, I'm not sure I understand the question again, but here is a more detailed description of what is happening in the line which I think is causing confusion:
`int b = tr_out[2];`

In the code above, the `operator[]` from `transform_iterator` will be invoked, resulting in an the transform `_Unary_Out<lambda>` to be applied to the `int&` at `a[2]`. `_Unary_Out` wraps the `int&` in an object of type `transform_output_ref_wrapper<int, lambda>` and returns it.  Then the "`=`" will trigger a conversion operator from `transform_output_ref_wrapper<int, lambda>` to `int&`, where the `int&` member variable within `transform_output_ref_wrapper<int, lambda>` is returned and copied into `b`.